### PR TITLE
Fogl 1658 - Update foglamp-south-am2315

### PR DESCRIPTION
**Pull request by amarendra-dianomic**

### CODE REVIEW

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 1]
```diff
@@ -28,6 +28,19 @@
          'type': 'string',
          'default': 'am2315'
     },
+    'assetPrefix': {
+        'description': 'Asset prefix',
+        'type': 'string',
+        'default': 'am2315_%M_',
+        'order': '2'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: it should be 1

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 2]
```diff
@@ -28,6 +28,19 @@
...
+    },
+    'i2cAddress': {
+        'description': 'I2C address in hex',
+        'type': 'string',
+        'default': '0x5C',
+        'order': '1'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: it should be 2

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 3]
```diff
@@ -28,6 +28,19 @@
          'type': 'string',
          'default': 'am2315'
     },
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: make plugin name  to load readonly

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 4]
```diff
@@ -161,11 +176,11 @@ def plugin_reconfigure(handle, new_config):
     diff = utils.get_diff(handle, new_config)
 
     # Plugin should re-initialize and restart if key configuration is changed
-    if 'pollInterval' in diff:
+    if 'pollInterval' in diff or 'i2cAddress' in diff:
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: what about assetPrefix? did we test restart?

**Review by praveen-garg — CHANGES_REQUESTED** [review]
remove _plugin_stop  it  neither has real code to execute nor a contract method.

**Review by Singhal-Vaibhav — CHANGES_REQUESTED** [review]
The plugin goes into a state of infinite loop when i2cAddress us updated. Syslog keep producing these errors:
```
ERROR: server: foglamp.services.south.server: Key error plugin am : 'bus'#012Traceback (most recent call last):#012  File "/home/pi/FogLAMP/python/foglamp/services/south/server.py", line 167, in _exec_plugin_poll#012    data = self._plugin.plugin_poll(self._plugin_handle)#012  File "/home/pi/FogLAMP/python/foglamp/plugins/south/am2315/am2315.py", line 103, in plugin_poll#012    bus = handle["bus"]#012KeyError: 'bus'
```

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 5]
```diff
@@ -161,26 +177,15 @@ def plugin_reconfigure(handle, new_config):
...
         new_handle = copy.deepcopy(new_config)
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/am2315/am2315.py]:
```
plugin_shutdown(handle)
new_handle = plugin_init(new_config)
new_handle['restart'] = 'yes'
_LOGGER.info(“Restarting AM2315 plugin due to change in configuration key [{}]“.format(‘, ’.join(diff)))
```

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 6]
```diff
@@ -126,7 +143,7 @@ def plugin_poll(handle):
             pass
         time_stamp = str(datetime.datetime.now(tz=datetime.timezone.utc))
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/am2315/am2315.py]: `utils.local_timestamp()`

**Review by Singhal-Vaibhav — DISMISSED** [review]

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# foglamp-south-am2315
```
Comment (praveen-garg) [inline README.rst]: rst fix needed

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 8]
```diff
@@ -26,13 +26,28 @@
...
          'description': 'AM2315 Poll Plugin',
          'type': 'string',
-         'default': 'am2315'
+         'default': 'am2315',
+         'readonly': 'true'
     },
+    'assetName': {
+        'description': 'Asset name',
+        'type': 'string',
+        'default': 'am2315/%M/',
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: do we want this `%M` to be replaced by i2c address? @ashwinscale  @billehunt  please confirm

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 9]
```diff
@@ -26,13 +26,28 @@
...
+        'order': '2'
+    },
+
     'pollInterval': {
         'description': 'The interval between poll calls to the South device poll routine expressed in milliseconds.',
         'type': 'integer',
-        'default': '5000'
-    }
+        'default': '5000',
+        'order': '2'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: let's add some min value ? I hope 0 is handled by south server

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 10]
```diff
@@ -161,26 +178,16 @@ def plugin_reconfigure(handle, new_config):
     diff = utils.get_diff(handle, new_config)
 
     # Plugin should re-initialize and restart if key configuration is changed
-    if 'pollInterval' in diff:
-        new_handle = copy.deepcopy(new_config)
-        new_handle['restart'] = 'no'
+    if 'pollInterval' in diff or 'i2cAddress' in diff:
+        plugin_shutdown()
+        new_handle = plugin_init(new_config)
+        new_handle['restart'] = 'yes'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/am2315/am2315.py]: do we need to restart for `pollInterval`

**Review by praveen-garg — CHANGES_REQUESTED** [review]

**Comment by amarendra-dianomic** [comment]
@Singhal-Vaibhav #3 merged into this PR. Pl check.

**Review by praveen-garg — DISMISSED** [review]
There are few comments which are not addressed. let's handle them in next PR with preinst script removal!

File: python/foglamp/plugins/south/am2315/am2315.py [hunk 11]
```diff
@@ -26,16 +27,32 @@
...
     },
+
     'pollInterval': {
         'description': 'The interval between poll calls to the South device poll routine expressed in milliseconds.',
         'type': 'integer',
-        'default': '5000'
-    }
+        'default': '5000',
+        'order': '2',
+        'minimum': 1000
```
Comment (Singhal-Vaibhav) [inline python/foglamp/plugins/south/am2315/am2315.py]: Should be a string

**Review by Singhal-Vaibhav — CHANGES_REQUESTED** [review]

**Comment by amarendra-dianomic** [comment]
@Singhal-Vaibhav minimum fixed. Pl check.

**Review by Singhal-Vaibhav — APPROVED** [review]
Code wise look good

**Pull request merged**